pkg/controller/controller_instance.go

```diff
@@ -236,139 +234,65 @@ func (c *controller) reconcileServiceInstanceDelete(instance *v1beta1.ServiceIns
 
 	glog.V(4).Info(pcb.Message("Processing deleting event"))
 
-	// Determine if any credentials exist for this instance.  We don't want to
-	// delete the instance if there are any associated creds
-	credentialsLister := c.bindingLister.ServiceBindings(instance.Namespace)
+	if instance.Status.OperationStartTime != nil &&
+		c.reconciliationRetryDurationExceeded(instance.Status.OperationStartTime) {
 
-	selector := labels.NewSelector()
-	credentialsList, err := credentialsLister.List(selector)
+		return c.updateServiceInstanceDeprovisionFailure(
+			instance,
+			errorReconciliationRetryTimeoutReason,
+			"Stopping reconciliation retries because too much time has elapsed",
+		)
+	}
+
+	// Determine if any bindings exist for this instance.  We don't want to
+	// delete the instance if there are any associated.
+	hasExistingBindings, err := c.serviceInstanceHasExistingBindings(instance)
 	if err != nil {
 		return err
 	}
-	for _, credentials := range credentialsList {
-		if instance.Name == credentials.Spec.ServiceInstanceRef.Name {
-
-			// found credentials, block the deletion until they are removed
-			clone, err := api.Scheme.DeepCopy(instance)
-			if err != nil {
-				return err
-			}
-			toUpdate := clone.(*v1beta1.ServiceInstance)
-
-			s := "Delete instance blocked by existing ServiceBindings associated with this instance.  All credentials must be removed first"
-			glog.Warning(pcb.Message(s))
-			c.recorder.Event(instance, corev1.EventTypeWarning, errorDeprovisionBlockedByCredentialsReason, s)
-
-			setServiceInstanceCondition(
-				toUpdate,
-				v1beta1.ServiceInstanceConditionReady,
-				v1beta1.ConditionFalse,
-				errorDeprovisionBlockedByCredentialsReason,
-				"Delete instance failed. "+s)
-			if _, err = c.updateServiceInstanceStatus(toUpdate); err != nil {
-				return err
-			}
-			return nil
-		}
+	if hasExistingBindings {
+		return c.updateServiceInstanceReconciliationError(
+			instance,
+			errorDeprovisionBlockedByCredentialsReason,
+			"All associated ServiceBindings must be removed before this ServiceInstance can be deleted",
+		)
 	}
 
-	// If the deprovisioning succeeded or is not needed, then clear out the finalizers
+	// If the deprovisioning succeeded or is not needed, then no need to
+	// make a request to the broker.
 	if instance.Status.DeprovisionStatus == v1beta1.ServiceInstanceDeprovisionStatusNotRequired ||
 		instance.Status.DeprovisionStatus == v1beta1.ServiceInstanceDeprovisionStatusSucceeded {
 
-		glog.V(5).Info(pcb.Message("Clearing catalog finalizer"))
-		clone, err := api.Scheme.DeepCopy(instance)
-		if err != nil {
-			return err
-		}
-		toUpdate := clone.(*v1beta1.ServiceInstance)
-		// Clear the finalizer
-		finalizers.Delete(finalizerToken)
-		toUpdate.Finalizers = finalizers.List()
-		if _, err := c.updateServiceInstanceStatus(toUpdate); err != nil {
-			return err
-		}
-		return nil
+		return c.updateServiceInstanceDeprovisionSuccess(instance)
```

Comment:
I don't think it is appropriate to call this function here. The function is going to do things that should not be done in the case where the Controller is not actually do a deprovision. The condition should not have the ready condition reason changed to deprovision successful, and the deprovision status should not be changed to Succeeded.

Reply:
Ah, I guess semantics-wise I can see us reserving the term "deprovision" for an actual deprovision request. Alright.